arch/mips/core/swap.S

```diff
@@ -0,0 +1,64 @@
+/*
+ * Copyright (c) 2017 Imagination Technologies Ltd.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <irq.h>
+#include <kernel_structs.h>
+#include <offsets_short.h>
+
+#include <mips/regdef.h>
+#include <mips/asm.h>
+#include <mips/cpu.h>
+#include <mips/hal.h>
+
+/* exports */
+GTEXT(__swap)
+GTEXT(_thread_entry_wrapper)
+
+/*
+ * unsigned int __swap(unsigned int key)
+ *
+ * Always called with interrupts locked
+ * key is stored in a0 register
+ */
+SECTION_FUNC(exception.other, __swap)
+
+	/* Set cause to trigger a software interrupt */
+    mfc0    k0, C0_CR
+    addi    k0, k0, 256
+    mtc0    k0, C0_CR
+
+    /* Enable interrupts */
+    mfc0 k0, C0_SR
+    or  k0, k0, a0
+    mtc0 k0, C0_SR
+    ehb
+
+    la k0, _kernel
+
+	/* Get pointer to _kernel.current */
+	lw k1, _kernel_offset_to_current(k0)
+
+	/* Load return value of __swap function in temp register v0 */
+	lw v0, _thread_offset_to_swap_return_value(k1)
+
+	jr ra
+
+/*
+ * void _thread_entry_wrapper(_thread_entry_t, void *, void *, void *)
+ */
+SECTION_FUNC(TEXT, _thread_entry_wrapper)
+	/*
+	 * _thread_entry_wrapper is called for every new thread upon the return
+	 * of __swap or ISR. Its address, as well as its input function
+	 * arguments thread_entry_t, void *, void *, void * are restored from
+	 * the thread stack (initialized via function _thread).
+	 * In this case, thread_entry_t, * void *, void * and void * are stored
+	 * in registers a0, a1, a2 and a3. These registers are used as arguments
+	 * to function _thread_entry. Hence, just call _thread_entry with
+	 * return address set to 0 to indicate a non-returning function call.
+	 */
+
```

Comment:
What mechanism pops the initial arguments off the stack? This function is a no-op. If this is all done transparently by the calling convention, why do we need this at all?